### ISSUE TITLE
Building pdf via LaTeX results in "TeX capacity exceeded" error with NumPy docs

### ISSUE THREAD

**Issue by rossbar** [sphinx-doc/sphinx#7723]
**Describe the bug**
When attempting to build a pdf of the NumPy reference guide, the `make all-pdf` step fails with the following:

```
! TeX capacity exceeded, sorry [input stack size=5000].

If you really absolutely need more capacity,                                    
you can ask a wizard to enlarge me.                                             
                                                                                
                                                                                
Here is how much of TeX's memory you used:                                      
 22864 strings out of 492167                                                    
 441790 string characters out of 6131558                                        
 817094 words of memory out of 5000000                                          
 21215 multiletter control sequences out of 15000+600000                        
 85482 words of font info for 135 fonts, out of 8000000 for 9000                
 1142 hyphenation exceptions out of 8191                                        
 5000i,26n,9989p,1538b,789s stack positions out of 5000i,500n,10000p,200000b,80000s
!  ==> Fatal error occurred, no output PDF file produced!
```

From analyzing the `.log`, it seems that the problem lies with LaTeX control sequences being injected into `\sphinxhref` commands. For example:

```tex
\sphinxhref{https://en.wikipedia.org/wiki/Two\textquotesingle{}s\_complement}{https://en.wikipedia.org/wiki/Two’s\_complement}
```

Manually replacing the `\textquotesingle{}` with `'` in the above block (and several other places where this occurs) allows the build to finish and the references to resolve correctly.

I'm not sure if this is a bug or if there is a configuration value I've missed that prevents this from happening.

**To Reproduce**

Minimal steps to set up an environment for building numpy docs from source:
```
$ git clone https://github.com/numpy/numpy.git && cd numpy
$ pip install cython
$ pip install -e .
$ pip install -r doc_requirements.txt
$ cd doc
$ git submodule update --init
```
Build pdf of refguide:
```
$ make latex
$ make -C build/latex all-pdf
```

**Expected behavior**
Build completes successfully with some warnings about unicode characters (unrelated to this issue)

**Your project**
https://github.com/numpy/numpy

**Environment info**
- OS: Arch Linux - 5.6.14-arch1-1
- Python version: 3.8.3
- Sphinx version: 2.4.4
- Sphinx extensions:  [                                                                  
    'sphinx.ext.autodoc',                                                       
    'numpydoc',                                                                 
    'sphinx.ext.intersphinx',                                                   
    'sphinx.ext.coverage',                                                      
    'sphinx.ext.doctest',                                                       
    'sphinx.ext.autosummary',                                                   
    'sphinx.ext.graphviz',                                                      
    'sphinx.ext.ifconfig',                                                      
    'matplotlib.sphinxext.plot_directive',                                      
    'IPython.sphinxext.ipython_console_highlighting',                           
    'IPython.sphinxext.ipython_directive',                                      
    'sphinx.ext.imgmath',                                                       
]
- Extra tools: Latexmk v. 4.65

**Additional context**

- numpy/numpy#15155

**Comment by rossbar** [comment]
Note: this problem is avoided with `xelatex` (`latex_engine='xelatex'`)

**Comment by shizhe1212** [comment]
Hi, I have the same error when I cite a Wikipedia webpage, could someone help me with this, please. And when I cite other url link, the error does not occur.

**Comment by tk0miya** [comment]
>\sphinxhref{https://en.wikipedia.org/wiki/Two\textquotesingle{}s\_complement}{https://en.wikipedia.org/wiki/Two’s\_complement}

I think this is not a valid URL. Please do URL escaping for this.

Note: I thought Sphinx escapes it automatically on building. But it causes double-escaping for other URLs. So I think it is the responsibility of users. It might be better to emit a warning if unexpected characters are detected on URL in runtime.

**Comment by rossbar** [comment]
Thanks @tk0miya for the insight. Note from the comment above that the `xelatex` engine doesn't have this problem. The link from the source rST file appears to be valid (no unicode characters nor any html characters that need escaping). I'm not sure what stage in the conversion the problem is being introduced. In any case, +1 on a warning. That would make this much more straightforward for users to identify and correct.

**Comment by tk0miya** [comment]
Finally, I added a code for skipping the escape of single quotes. It will be released as 3.1.0.
Thank you for reporting!
